src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/Utilities/CompareHelper`1.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Razor.Utilities;
+
+/// <summary>
+///  Helper that avoids creating an <see cref="IComparer{T}"/> until its needed.
+/// </summary>
+internal readonly ref struct CompareHelper<T>
+{
+    private readonly IComparer<T> _comparer;
+    private readonly Comparison<T> _comparison;
+    private readonly bool _comparerSpecified;
+    private readonly bool _useComparer;
+    private readonly bool _descending;
+
+    public CompareHelper(IComparer<T>? comparer, bool descending)
+    {
+        _comparerSpecified = comparer is not null;
+        _comparer = comparer ?? Comparer<T>.Default;
```

Comment:
Personal opinion, but I'd consider moving the logic for setting this comparer fully into this method, and removing `_comparerSpecified` and `_useComparer`. It would simplify the switch below a lot, and mean almost all decisions are simply based on whether `_comparer` or `_comparison` is set.

I might be wrong though, and it could be just as complicated, so take the suggestion as you will.

Reply:
Should be much simpler now.